cpp/src/arrow/util/future.cc

```diff
@@ -272,8 +315,8 @@ class ConcreteFutureImpl : public FutureImpl {
     //
     // In fact, it is important not to hold the locks because the callback
     // may be slow or do its own locking on other resources
-    for (auto&& callback : callbacks_) {
-      std::move(callback)();
+    for (auto& callback_record : callbacks_) {
+      RunOrScheduleCallback(callback_record, /*from_unfinished=*/true);
```

Comment:
Shouldn't `from_unfinished` be false?

Reply:
Hmm, `from_unfinished` is supposed to mean "was the callback added when the future was unfinished" but I can see how that is vague.  I could just remove this option entirely.  The `TryAddCallback` function already gives you this capability (it is used in `Transfer`) and so it isn't used anywhere.  Any strong opinion?

Reply:
On the bright side, if we remove `IF_UNFINISHED` then we can change `ShouldSchedule` to a `bool`.

Reply:
I think `IfUnfinished` is ok, perhaps the `from_unfinished` variable needs better naming or an explanatory comment though :-)

Reply:
Ok, I renamed it to `in_add_callback` (and inverted the logic).  Let me know if that is more clear.